toolchain/lexer/tokenized_buffer.cpp

```diff
@@ -907,23 +907,37 @@ auto TokenizedBuffer::SourceBufferLocationTranslator::GetLocation(
   int line_number = line_it - buffer_->line_infos_.begin();
   int column_number = offset - line_it->start;
 
+  llvm::StringRef line;
+
   // We might still be lexing the last line. If so, check to see if there are
   // any newline characters between the position we've finished lexing up to
   // and the given location.
   if (incomplete_line_info && column_number > *last_line_lexed_to_column_) {
     column_number = *last_line_lexed_to_column_;
+    int64_t start = line_it->start;
     for (int64_t i = line_it->start + *last_line_lexed_to_column_; i != offset;
          ++i) {
       if (buffer_->source_->text()[i] == '\n') {
+        start = i;
         ++line_number;
         column_number = 0;
       } else {
         ++column_number;
       }
     }
+    line = buffer_->source_->text().substr(start).take_until(
+        [](char c) { return c == '\n'; });
+  } else if (line_it->length < 0) {
+    line =
+        buffer_->source_->text().substr(line_it->start).take_until([](char c) {
+          return c == '\n';
+        });
+  } else {
+    line = buffer_->source_->text().substr(line_it->start, line_it->length);
   }
```

Comment:
```suggestion
  // We might still be lexing the last line. If so, check to see if there are
  // any newline characters between the position we've finished lexing up to
  // and the given location.
  int64_t start = line_it->start;
  if (incomplete_line_info && column_number > *last_line_lexed_to_column_) {
    column_number = *last_line_lexed_to_column_;
    for (int64_t i = start + *last_line_lexed_to_column_; i != offset;
         ++i) {
      if (buffer_->source_->text()[i] == '\n') {
        start = i;
        ++line_number;
        column_number = 0;
      } else {
        ++column_number;
      }
    }
  }
  
  int64_t length = line_it->length;
  if (length < 0) {
    length = buffer_->source_->text().find_first_of('\n', start);
  }

  llvm::StringRef line = buffer_->source_->text().substr(start, length);
```

I think we can simplify this a little.

Reply:
I have a different change in 3011. I'm hesitant here to rely on how if no `\n` is found then npos results in substr going to the end, I think that can be a little confusing.

Reply:
Sidenote -- the length calculation on 930 is incorrect, that's the offset from the original line start, which can be changed on line 919